File: java/algtbox/src/MaxPairwiseProduct.java
import java.util.*;
import java.io.*;

public class MaxPairwiseProduct {
    static long getMaxPairwiseProduct(int[] numbers) {
        int n = numbers.length;
        int top1 = 0, top2 = 0;

        for (int i = 0; i < n; ++i) {
           if (numbers[i] > top1)
           {
               top2 = top1;
               top1 = numbers[i];
           }
           else if (numbers[i] > top2)
           {
               top2 = numbers[i];
           }
        }

        long result = (long)top1 * top2;
        return result;
    }

    public static void main(String[] args) {

//        int[] numbers = { 100000, 90000 };
        FastScanner scanner = new FastScanner(System.in);
        int n = scanner.nextInt();
        int[] numbers = new int[n];
        for (int i = 0; i < n; i++) {
            numbers[i] = scanner.nextInt();
        }
        System.out.println(getMaxPairwiseProduct(numbers));
    }

    static class FastScanner {
        BufferedReader br;
        StringTokenizer st;

        FastScanner(InputStream stream) {
            try {
                br = new BufferedReader(new InputStreamReader(stream));
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        String next() {
            while (st == null || !st.hasMoreTokens()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }
    }

}

File: alg/scala/src/MaxPairwiseProduct.scala
import java.util.Scanner


object MaxPairwiseProduct {
  def main(args: Array[String]) = {
    val s = new Scanner(System.in);

    val n = s.nextInt();
    val numbers : Array[Int] = new Array[Int](n);
    for (i <- 0 to n-1) numbers(i) = s.nextInt()
    println( getMaxPairwiseProduct(numbers) );
  }

  def getMaxPairwiseProduct (numbers: Array[Int]) : Long =
  {
    var top1 : Long = 0;
    var top2 : Long = 0;

    for (n <- numbers)
    {
      if (n > top1)
      {
        top2 = top1
        top1 = n
      }
      else if (n > top2)
      {
        top2 = n
      }
    }

    top1*top2
  }

}
